# Importing and Processing Bee Data into Pandas DataFrames and CSV Files

Bee-related data is sourced from various datasets stored in both .xlsx and .csv formats. Some .xlsx files contain multiple sheets, each representing data for a specific year.

This notebook aims to import and process each data file, saving the processed data as .csv files. Subsequently, the saved files will undergo cleaning and exploratory data analysis (EDA).

In [1]:
# imports
import pandas as pd
import os
import pandas as pd

### Creating a directory using os library to store processed data frames

*citation:* [Using `os` to make directories]('https://www.w3schools.com/python/ref_os_makedirs.asp')

In [2]:
directory = '../Project_4/Data_Bees/processed_dfs/'
os.makedirs(directory, exist_ok = True)

---
## 1 ) Function to import and process data from all sheets/years in `BIP Bee Colony Loss Clean.xlsx`
*citation:* [Looking up sheet names in .xlsx notebook]('https://stackoverflow.com/questions/17977540/pandas-looking-up-the-list-of-sheets-in-an-excel-file')

In [3]:
# Creating the function
def process_data(file): # year data is in the format 2007-08
    
    file = str(file) 
    dfs = [] #store each data frame, for concatenation

    file_for_import = f'../Project_4/Data_Bees/raw_bee_data/{file}.xlsx'
    sheet_names = pd.ExcelFile(file_for_import).sheet_names #see citation
    
    for sheet_name in sheet_names:
        df = pd.read_excel(file_for_import, sheet_name = sheet_name)
        df['year'] = str(sheet_name[0:4]) #make a column for the year
        dfs.append(df) #append df to the list for concatenation

    #concatenate
    df_combo = pd.concat(dfs, ignore_index = True)

    return df_combo

# saving the output df as a variable
colony_loss_df = process_data('BIP Bee Colony Loss Clean')

In [4]:
colony_loss_df.head()

State Total Winter All Loss  Beekeepers Beekeepers Exclusive to State  \
0    Maryland                  7.6%          14                          100%   
1  Washington                 13.7%           5                            0%   
2  New Jersey                 15.1%          15                           80%   
3    Arkansas                 17.4%          20                          100%   
4       Maine                   18%           6                         16.7%   

   Colonies Colonies Exclusive to State  year  
0      4013                        100%  2007  
1     21870                          0%  2007  
2     22622                         12%  2007  
3     16955                        100%  2007  
4     45937                        0.1%  2007

In [5]:
colony_loss_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 557 entries, 0 to 556
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   State                          556 non-null    object
 1   Total Winter All Loss          557 non-null    object
 2   Beekeepers                     557 non-null    int64 
 3   Beekeepers Exclusive to State  557 non-null    object
 4   Colonies                       557 non-null    int64 
 5   Colonies Exclusive to State    557 non-null    object
 6   year                           557 non-null    object
dtypes: int64(2), object(5)
memory usage: 30.6+ KB


---
The following columns are stored as 'object' instead of 'float':
'Total Winter All Loss', 'Beekeepers Exclusive to State', 'Colonies Exclusive to State', 'year'.

In [6]:
cols_to_change = ['Total Winter All Loss', 
                  'Beekeepers Exclusive to State', 
                  'Colonies Exclusive to State', 
                  'year']

def turn_to_float(df):
    for col in cols_to_change:
        if '%' in str(df[col]):
            df[col] = df[col].str.replace('%', '').astype(float)/100
        else:
            df[col] = df[col].astype(float)
    return df

turn_to_float(colony_loss_df)

State  Total Winter All Loss  Beekeepers  \
0      Maryland                  0.076          14   
1    Washington                  0.137           5   
2    New Jersey                  0.151          15   
3      Arkansas                  0.174          20   
4         Maine                  0.180           6   
..          ...                    ...         ...   
552    Illinois                  0.546         108   
553         NaN                  0.560           5   
554        Iowa                  0.618          83   
555   Louisiana                  0.628           8   
556      Nevada                  0.655          15   

     Beekeepers Exclusive to State  Colonies  Colonies Exclusive to State  \
0                            1.000      4013                        1.000   
1                            0.000     21870                        0.000   
2                            0.800     22622                        0.120   
3                            1.000     16955                        1.000   
4                            0.167     45937                        0.001   
..                             ...       ...                          ...   
552                          0.963       951                        0.965   
553                          0.800       352                        0.034   
554                          0.964       857                        0.972   
555                          0.875     13102                        0.007   
556                          1.000       119                        1.000   

       year  
0    2007.0  
1    2007.0  
2    2007.0  
3    2007.0  
4    2007.0  
..      ...  
552  2018.0  
553  2018.0  
554  2018.0  
555  2018.0  
556  2018.0  

[557 rows x 7 columns]

In [7]:
# saving to .csv
colony_loss_df.to_csv(directory + 'bip_colony_loss_processed.csv')

---
## 2) Import and process data from `Bee Colony Census Data by State.csv`

In [8]:
df_2 = pd.read_csv('./Data_Bees/raw_bee_data/Bee Colony Census data by State.csv')

In [9]:
df_2.describe()

Year  Week Ending  State ANSI  Ag District  Ag District Code  \
count   200.000000          0.0  200.000000          0.0               0.0   
mean   2004.500000          NaN   29.320000          NaN               NaN   
std       5.604198          NaN   15.662829          NaN               NaN   
min    1997.000000          NaN    1.000000          NaN               NaN   
25%    2000.750000          NaN   17.000000          NaN               NaN   
50%    2004.500000          NaN   29.500000          NaN               NaN   
75%    2008.250000          NaN   42.000000          NaN               NaN   
max    2012.000000          NaN   56.000000          NaN               NaN   

       County  County ANSI  Zip Code  Region  watershed_code  Watershed  \
count     0.0          0.0       0.0     0.0           200.0        0.0   
mean      NaN          NaN       NaN     NaN             0.0        NaN   
std       NaN          NaN       NaN     NaN             0.0        NaN   
min       NaN          NaN       NaN     NaN             0.0        NaN   
25%       NaN          NaN       NaN     NaN             0.0        NaN   
50%       NaN          NaN       NaN     NaN             0.0        NaN   
75%       NaN          NaN       NaN     NaN             0.0        NaN   
max       NaN          NaN       NaN     NaN             0.0        NaN   

          CV (%)  
count  50.000000  
mean   21.090000  
std    17.735198  
min     2.700000  
25%     6.775000  
50%    14.150000  
75%    33.650000  
max    74.900000

In [10]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Program           200 non-null    object 
 1   Year              200 non-null    int64  
 2   Period            200 non-null    object 
 3   Week Ending       0 non-null      float64
 4   Geo Level         200 non-null    object 
 5   State             200 non-null    object 
 6   State ANSI        200 non-null    int64  
 7   Ag District       0 non-null      float64
 8   Ag District Code  0 non-null      float64
 9   County            0 non-null      float64
 10  County ANSI       0 non-null      float64
 11  Zip Code          0 non-null      float64
 12  Region            0 non-null      float64
 13  watershed_code    200 non-null    int64  
 14  Watershed         0 non-null      float64
 15  Commodity         200 non-null    object 
 16  Data Item         200 non-null    object 
 1

In [11]:
df_2.isna().sum()

Program               0
Year                  0
Period                0
Week Ending         200
Geo Level             0
State                 0
State ANSI            0
Ag District         200
Ag District Code    200
County              200
County ANSI         200
Zip Code            200
Region              200
watershed_code        0
Watershed           200
Commodity             0
Data Item             0
Domain                0
Domain Category       0
Value                 0
CV (%)              150
dtype: int64

---
#### Dropping Columns

There are several columns we won't use in our analysis. Additionally, several columns are mosty blank.

* blank columns = 'Week Ending', 'Ag District', 'Ag District Code', 'County', 'County ANSI', 'Zip Code', 'Region', 'Watershed'

* columns with all the same value:
    -'Domain Category' = 'NOT SPECIFIED'
    -'Domain' = 'TOTAL'
    -'Data Item' = 'HONEY, BEE COLONY'
    -'Commodity' = 'HONEY'
    -'watershed_code' =0
    -'Program' = 'Census'
    -'Period' = 'END OF DEC' 

* unusable column  =  CV(%) is missing 150 values out of 200. CV(%) is the Coefficient of variation. Available for the 2012 Census of Agriculture only. 

In [12]:
df_2.drop(columns = ['Week Ending', 'Ag District', 'Ag District Code', 'County', 'County ANSI', 'Zip Code', 'Region', 'Domain Category', 'Domain', 'Data Item', 'Commodity', 'watershed_code', 'State ANSI', 'Geo Level', 'Period', 'Program', 'Watershed', 'CV (%)'], inplace = True)

In [13]:
df_2.head()


Year       State    Value
0  2012     ALABAMA   11,628
1  2012      ALASKA      546
2  2012     ARIZONA   58,461
3  2012    ARKANSAS   23,259
4  2012  CALIFORNIA  945,589

In [14]:
df_2[df_2['Year'] == 2008]

Empty DataFrame
Columns: [Year, State, Value]
Index: []

**OBSERVATION:** There is data for 4 years only

In [15]:
# The 'value' column in this dataset will be explicitly defined for future merging or concatenation with other dataframes.
df_2['census_value'] = df_2['Value']

In [16]:
# saving to .csv
df_2.to_csv(directory + 'state_census_processed.csv')

---
## Importing Bee Colony Survey Data by State


In [17]:
df_3 = pd.read_csv('./Data_Bees/raw_bee_data/Bee Colony Survey Data by State.csv')
df_3.head()

Year        Period  Week Ending       State  State ANSI  Watershed  \
0  2017  JAN THRU MAR          NaN     ALABAMA           1        NaN   
1  2017  JAN THRU MAR          NaN     ARIZONA           4        NaN   
2  2017  JAN THRU MAR          NaN    ARKANSAS           5        NaN   
3  2017  JAN THRU MAR          NaN  CALIFORNIA           6        NaN   
4  2017  JAN THRU MAR          NaN    COLORADO           8        NaN   

          Data Item    Value  CV (%)  
0  ADDED & REPLACED      570     NaN  
1  ADDED & REPLACED    2,900     NaN  
2  ADDED & REPLACED      430     NaN  
3  ADDED & REPLACED  215,000     NaN  
4  ADDED & REPLACED      100     NaN

In [18]:
df_3.drop(columns = ['CV (%)', 'Watershed', 'Week Ending', 'State ANSI'], inplace = True)

In [19]:
df_3.sort_values(by= ['State'])

Year          Period    State         Data Item   Value
0     2017    JAN THRU MAR  ALABAMA  ADDED & REPLACED     570
2307  2011  MARKETING YEAR  ALABAMA         INVENTORY   9,000
405   2015    OCT THRU DEC  ALABAMA  ADDED & REPLACED      80
2676  2002  MARKETING YEAR  ALABAMA         INVENTORY  12,000
1528  2015    JUL THRU SEP  ALABAMA     LOSS, DEADOUT   1,400
...    ...             ...      ...               ...     ...
1617  2015    OCT THRU DEC  WYOMING     LOSS, DEADOUT   1,300
3155  1992  MARKETING YEAR  WYOMING         INVENTORY  41,000
584   2016    JAN THRU MAR  WYOMING    INVENTORY, MAX   6,500
1527  2015    APR THRU JUN  WYOMING     LOSS, DEADOUT   3,200
3395  1987  MARKETING YEAR  WYOMING         INVENTORY  39,000

[3396 rows x 5 columns]

In [20]:
df_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3396 entries, 0 to 3395
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Year       3396 non-null   int64 
 1   Period     3396 non-null   object
 2   State      3396 non-null   object
 3   Data Item  3396 non-null   object
 4   Value      3396 non-null   object
dtypes: int64(1), object(4)
memory usage: 132.8+ KB


**The 'Value' is an object, it needs to be an int or float
Also, 'Value' has comas in it... which could be problematic**


In [21]:
df_3['Value'] = df_3['Value'].str.replace(',', '').astype(int) # needed .str for .replace to work

In [22]:
# I need to combine the periods of months in the 'Period' column to be yearly data
# First, deconstruct the Period Data, into a start and end date, to then use .to_datetime()

def convert_period_to_dates(entry):
    year = entry['Year'] # --> using entry['column'], will make the function go through the rows
    period = entry['Period']

    if 'THRU' in period:
        start_mo, end_mo = period.split('THRU')
        start_date = f'{start_mo} {year}'
        end_date = f'{end_mo} {year}'  
    else:
        start_date = period + f'{year}'
        end_date = period + f'{year}'
        
    return pd.Series([start_date, end_date]) 

# Create two new columsn for start_date and end_date Apply the function to data frame
df_3[['date_start', 'date_end']] = df_3.apply(convert_period_to_dates, axis = 1)

df_3.head()

Year        Period       State         Data Item   Value date_start  \
0  2017  JAN THRU MAR     ALABAMA  ADDED & REPLACED     570  JAN  2017   
1  2017  JAN THRU MAR     ARIZONA  ADDED & REPLACED    2900  JAN  2017   
2  2017  JAN THRU MAR    ARKANSAS  ADDED & REPLACED     430  JAN  2017   
3  2017  JAN THRU MAR  CALIFORNIA  ADDED & REPLACED  215000  JAN  2017   
4  2017  JAN THRU MAR    COLORADO  ADDED & REPLACED     100  JAN  2017   

    date_end  
0   MAR 2017  
1   MAR 2017  
2   MAR 2017  
3   MAR 2017  
4   MAR 2017

In [23]:
# dropping the 'period col'
df_3 = df_3.drop(columns = ['Period'])
df_3.head()

Year       State         Data Item   Value date_start   date_end
0  2017     ALABAMA  ADDED & REPLACED     570  JAN  2017   MAR 2017
1  2017     ARIZONA  ADDED & REPLACED    2900  JAN  2017   MAR 2017
2  2017    ARKANSAS  ADDED & REPLACED     430  JAN  2017   MAR 2017
3  2017  CALIFORNIA  ADDED & REPLACED  215000  JAN  2017   MAR 2017
4  2017    COLORADO  ADDED & REPLACED     100  JAN  2017   MAR 2017

In [24]:
df_3.columns

Index(['Year', 'State', 'Data Item', 'Value', 'date_start', 'date_end'], dtype='object')

---
*citations:* [aggregating columns]('https://stackoverflow.com/questions/49783178/keep-other-columns-when-using-sum-with-groupby')

In [25]:
# Group the DataFrame by 'Year', 'State', and 'Data Item', then sum the 'Value' column
df_3 = df_3.groupby(['Year', 'State', 'Data Item'])['Value'].sum().reset_index()

In [26]:
df_3 = df_3[df_3['Data Item'] == 'INVENTORY']
df_3.head()

Year       State  Data Item   Value
0  1987     ALABAMA  INVENTORY   46000
1  1987     ARIZONA  INVENTORY   80000
2  1987    ARKANSAS  INVENTORY   29000
3  1987  CALIFORNIA  INVENTORY  540000
4  1987    COLORADO  INVENTORY   44000

In [27]:
df_3['State'] = df_3['State'].str.lower()

In [28]:
df_3.drop(columns = ['Data Item'], inplace = True)
df_3.head()

Year       State   Value
0  1987     alabama   46000
1  1987     arizona   80000
2  1987    arkansas   29000
3  1987  california  540000
4  1987    colorado   44000

In [29]:
# saving to .csv
df_3.to_csv(directory + 'state_survey_processed.csv')